### PR TITLE
Allow user decide whether to automatically generate Propagation-policy and its name when promote

### LINKED ISSUE #3429: Letting the user decide whether to automatically generate Propagation-policy when Promote
**What would you like to be added**:
Adding a flag like `--auto-generate-propagation-policy=true` or `propagation-policy-name=*` to command
```shell
karmadactl promote <RESOURCE_TYPE> <RESOURCE_NAME> -n <NAME_SPACE> -C <CLUSTER_NAME>
```

**Why is this needed**:
My story is: I want to do a migration of non-federal applications on the legacy cluster to Karmada Control Plane, then propagate the application to other member clusters. I used the `promote` command and noticed it creates `propagation-policy` with a suffix in its succession. So I need to get the `pp` and edit it manually(which requires a little more work in a scription way).

if the `--auto-generate-propagation-policy=true` flag is supported, I can manually create the `pp` with my other specific requests rather than the default one.

if the `propagation-policy-name=*` flag is supported I don't have to distinguish which is the automatically generated `pp` and then edit it. And it will be easier to get all processes scripted.

Any of these flags will meet my needs here.

Since I have to edit the `propagation-policy` anyway, I think why not let me decide whether to create a default one or how -- decide its name, even more.

Ps. Isn't it more reasonable to let the user create the `propagation-policy` on their thoughts instead of `promote`? Or the `"Promote"` **IS** designed to imply the propagation part? What's your opinion?

### PR DESCRIPTION
**Pull request by LronDC**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
It's reasonable for users to let them decide whether to automatically create a policy and the name of auto-created PropagationPolicy.
**Which issue(s) this PR fixes**:
Fixes #3429 

**Special notes for your reviewer**:

My first time contributing to Karmada so any advice and guidance are expected.
Is a test case needed here for this change?
And I think another PR of the document's change should be opened.
**Does this PR introduce a user-facing change?**:

```release-note
`karmadactl`: Introduced `--auto-create-policy` and `--policy-name` flags for `promote` command to customize the policy during the promotion.
```

### CODE REVIEW

File: pkg/karmadactl/promote/promote.go
Comment (jwcesign) [inline pkg/karmadactl/promote/promote.go:118]: `AutoCreatePropagationPolicy` is better?

File: pkg/karmadactl/promote/promote.go
Comment (jwcesign) [inline pkg/karmadactl/promote/promote.go:123]: If this policy already exists, what will happen?

File: pkg/karmadactl/promote/promote.go
Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:118]: Sure, will do

File: pkg/karmadactl/promote/promote.go
Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:123]: before building one policy it will [check] if the name exists (https://github.com/karmada-io/karmada/pull/3494/files/1be3cf421633636a171dc4129069a249fd0569e0#diff-6d515eefa1bc0f6f344292ee9f555c0005bc04d0905307bb67357cb234b43ca6R402)

**Comment by jwcesign** [comment]
Hi, @LronDC , please squash the two commits together.
Other lgtm, cc @lonelyCZ for checking

**Comment by LronDC** [comment]
> Hi, @LronDC , please squash the two commits together. Other lgtm, cc @lonelyCZ for checking

Okay, may I wait for others' opinions? I will squash the commits after all changes be done.

**Comment by jwcesign** [comment]
> I will squash the commits after all changes be done.

Sure.

File: pkg/karmadactl/promote/promote.go
Comment (lonelyCZ) [inline pkg/karmadactl/promote/promote.go:139]: Could we only specify `propagation-policy-name`？When the name is nil, auto-gen the name, otherwise using the specific name.

What do you think? @LronDC @jwcesign

Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:139]: You mean leave the pp creation automatically, just let the user decide its name, do I get it correctly?

Comment (jwcesign) [reply pkg/karmadactl/promote/promote.go:139]: Do you mean that there is only one option for `propagation-policy-name`? If that's the case, then the policy will be automatically created, which doesn't solve the issue of manual creation.

Comment (lonelyCZ) [reply pkg/karmadactl/promote/promote.go:139]: Got it, it has two phase.

I think it is ok.

File: pkg/karmadactl/promote/promote.go
Comment (lonelyCZ) [inline pkg/karmadactl/promote/promote.go:140]:
```suggestion
		"The name of a PropagationPolicy that automatically created by command. If not specified, a suffix will be added to the PropagationPolicy")
```

The suffix is not generated random. It is a specific hash code.

File: pkg/karmadactl/promote/promote.go
Comment (lonelyCZ) [inline pkg/karmadactl/promote/promote.go:150]: Please delete this blank line.

File: pkg/karmadactl/promote/promote.go
Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:140]: Yes, you are right, nice catch! Thank you.

**Comment by lonelyCZ** [comment]
Hi, @RainbowMango , please help to trigger the CI.

**Comment by RainbowMango** [comment]
Done.

**Comment by lonelyCZ** [comment]
Hi, @LronDC , have you tested it in your env? Could you please share the test result?

**Comment by LronDC** [comment]
> Hi, @LronDC , have you tested it in your env? Could you please share the test result?

Yes. I tested it and it worked as expected. As you can see I failed last CI workflow on `e2e test` and I was tring to make it right, but bumped into some local env issues. I think I could figure it out today or tomorrow. Sorry for the late.

**Comment by LronDC** [comment]
@RainbowMango  Hi, I failed the last CI on the e2e test, I've made some changes and tested it in my local. It worked fine and passed the e2e test now. Could you please help to trigger the CI again?

**Comment by RainbowMango** [comment]
Done.

**Comment by lonelyCZ** [comment]
I just tested it in my env, it also works fine as expected!

Very thanks @LronDC !

/approve

**Review by XiShanYongYe-Chang — COMMENTED** [review]
For e2e part, LGTM

Because cross-file directories are involved, approve of @RainbowMango may be required.
/cc @RainbowMango

File: pkg/karmadactl/promote/promote.go
Comment (RainbowMango) [inline pkg/karmadactl/promote/promote.go:139]: Hold on, I'm looking at the flags.

File: pkg/karmadactl/promote/promote.go
Comment (RainbowMango) [inline pkg/karmadactl/promote/promote.go:140]: The flag `auto-create-propagation-policy` might be misleading, it would be better to describe it more clearly.

My suggestion:
```suggestion
	flags.BoolVar(&o.AutoCreatePolicy, "auto-create-policy", true,
		"Automatically create a PropagationPolicy for namespace-scoped resources or create a ClusterPropagationPolicy for cluster-scoped resources.")
	flags.StringVar(&o.PolicyName, "policy-name", "",
		"The name of the PropagationPolicy(or ClusterPropagationPolicy) that is automatically created after promotion. If not specified, the name will be the resource name with a random hash suffix.")
```

File: pkg/karmadactl/promote/promote.go
Comment (RainbowMango) [inline pkg/karmadactl/promote/promote.go:123]:
```suggestion
	// AutoCreatePolicy tells if the promote command should create the
	// PropagationPolicy(or ClusterPropagationPolicy) by default.
	// It defaults to true.
	AutoCreatePolicy bool
	
	// PolicyName is the name of the PropagationPolicy(or ClusterPropagationPolicy),
	// It defaults to the promoting resource name with a random hash suffix.
	// It will be ignored if AutoCreatePolicy is false.
	PolicyName string
```

**Review by RainbowMango — COMMENTED** [review]
Please @lonelyCZ take another look at the suggestions.

File: pkg/karmadactl/promote/promote.go
Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:140]: I think it's really more appropriate that include a description of namespace-scoped resources and cluster-scoped resources.

One thing. @lonelyCZ had mentioned that the default name suffix is not "random" but a specific hash code actually. Do we consider to add "random" word to the description?

Comment (lonelyCZ) [reply pkg/karmadactl/promote/promote.go:140]: How about `a specific hash suffix`?

Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:140]: > How about `a specific hash suffix`?
If we are on a agreement here I'm gonna make the change.:)

Comment (RainbowMango) [reply pkg/karmadactl/promote/promote.go:140]: `with a hash suffix that is generated by resource metadata`?

Comment (RainbowMango) [reply pkg/karmadactl/promote/promote.go:140]: no big deal. I mean random because from user's respective he/she can't know the hash value and doesn't need to know.

Comment (LronDC) [reply pkg/karmadactl/promote/promote.go:140]: Got it.

Comment (LronDC) [reply pkg/karmadactl/promote/promote.go]: @RainbowMango Change be done. Need one more trigging to CI, and review labels 😄

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Review by RainbowMango — COMMENTED** [review]
This PR was blocked by fake testing, after the retesting it will be get merged automatically.

I just updated the release note according to the newly implementation.

**Pull request merged**
